src/layer/LayerGroup.js

```diff
@@ -50,6 +52,26 @@ export const LayerGroup = Layer.extend({
 		return this;
 	},
 
+	_preventLayerGroupRecursion(layer) {
+		if (layer instanceof LayerGroup && layer.hasDescendant(this)) {
+			throw new Error('Cannot add a LayerGroup as a child if it is already included, directly or nested.');
+		}
+	},

```

Comment:
This is a huge breaking-change, I don't know what I should thing about it

Reply:
Ummm, you're right. Sorry, I didn't think about this specifically...

So, in this case, I think we'd want to go with the recursive solution. It would be a non-breaking change, and it would also solve both issues.